multigram/README.md

```diff
@@ -0,0 +1,96 @@
+# Multigram -- protocol negotiation and multiplexing over datagrams
+
+For multiplexing different protocols on the same datagram connection, multigram prepends a 1-byte header to every packet. This header represents an index in a table of protocols shared between both endpoints. This protocol table is negotiated by exchanging the intersection of the endpoint's supported protocols. The protocol table's size of 256 tuples can be increased by nesting multiple multigram headers.
```

Comment:
Mh yep we're trading off between a complication of header parsing, and a complication of protocol semantics. We should decide whether we favor performance (int8 field and nested multigrams) over a simpler protocol (varint and no nesting).

We should figure out exactly *how complicated* nesting is (make experiments), and also *how heavy* parsing a varint is vs. an int8 (benchmarks). My gut feeling is that we should probably go with an int8, because we'll have to do this for *every* packet, and multiple times per packet in many use cases.

Not-so-appealing idea: a varint degrades to an int8 nicely, so there might a way to leave it up to implementations.